### PR TITLE
Close #9 and close #10

### LINKED ISSUE #9: Allow local credentials overiwriting global default credentials
目前 paperclip-storage-aliyun 每次建立 OSS 连接都使用了默认配置：

```ruby
oss_connection ||= ::Aliyun::Connection.new Paperclip::Attachment.default_options[:aliyun]
```

我们可以更灵活地支持在模型定义 `has_attached_file` 时覆盖默认配置，例如：

```ruby
#  paperclip-storage-aliyun/lib/paperclip/storage/aliyun.rb
oss_connection ||= ::Aliyun::Connection.new Paperclip::Attachment.default_options[:aliyun].merge(local_options)
```

这样我们就可以有

```ruby
has_attached_file :download, storage: :aliyun, aliyun: { bucket: 'another-place'}
has_attached_file :secret, storage: :aliyun, aliyun: { access_id: 'another-id', access_key: 'another-key', bucket: 'secret-place'}
```

### LINKED ISSUE #10: Provide internal_url and external_url at the same time
默认的 url 应当跟随配置中的 :internal 选项，但是实际中我们可能同时需要内部地址和外部地址，此时可以考虑给 `Paperclip::Attachment` 添加两个新方法：`internal_url` & `external_url`。

上传时，按服务器位置配置 :internal 选项，按 object key 命名规则配置  :path 即可自动生成 `url`, `internal_url` 和 `external_url`。使用时，可以有：

```ruby
# in photos_controller.rb
@faces = detect_faces(@photo.internal_url) # 使用 internal_url 以节约流量和加快速度
```

```haml
-# in views/photos/show.html.haml
= image_tag(@photo.external_url)
%caption There are #{@faces.count} faces in the above photo.
```

仅作建议：）

### PR DESCRIPTION
**Pull request by aidistan**
I have tried to keep old codes as many as possible...

Since there are several incompatible changes, I suggest to bump the version number to `0.1.0`.

### CODE REVIEW

**Comment by princemaple** [comment]
oh you are ahead of me again, fixes #13

File: lib/aliyun/connection.rb [hunk 1]
```diff
@@ -178,11 +180,11 @@ def get_bucket_path(path)
     # @param path [String] the path to retrieve the file on remote storage
     # @return [String] the expected full path, e.g. "http://martin-test.oss-cn-hangzhou.aliyuncs.com/oss-api.pdf"
     def path_to_url(path)
-      return path if path =~ /^https?:\/{2}/  # 已经是全路径
-      "http://#{fetch_file_host}/#{path}"
+      path =~ %r{^https?:\/{2}} ? path : "http://#{aliyun_upload_host}/#{path}"
```
Comment (princemaple) [inline lib/aliyun/connection.rb]: don't need to escape forward slash in `%r{}`

Comment (princemaple) [inline lib/aliyun/connection.rb]: also, `%r{^https?://}` is cleaner than using `/{2}`

File: lib/aliyun/data_center.rb [hunk 2]
```diff
@@ -1,20 +1,35 @@
...
+        fail InvalildDataCenter, "Unsupported Data Center #{options[:data_center]} Detected"
+      end
+
+      "#{data_center}#{options[:internal] ? '-internal' : ''}.aliyuncs.com"
+    end
+
+    def find_center(data_center)
+      return if /(oss|cn|us|ap|oss-cn)/.match(data_center)
+
+      AVAILABLE_DATA_CENTERS.each do |center|
```
Comment (princemaple) [inline lib/aliyun/data_center.rb]:
```ruby
matcher = Regexp.new(data_center) # instantiate once
AVAILABLE_DATA_CENTERS.find { |dc| matcher.match(dc) }
```

File: lib/paperclip/storage/aliyun.rb [hunk 3]
```diff
@@ -2,9 +2,34 @@ module Paperclip
...
+          :aliyun_external_url, :aliyun_alias_url
+        ].each do |url_style|
+          Paperclip.interpolates(url_style) do |attachment, style|
+            attachment.send(url_style, style)
+          end unless Paperclip::Interpolations.respond_to? url_style
+        end
+      end
+
+      def aliyun_upload_url(style = default_style)
+        "http://#{oss_connection.aliyun_upload_host}/#{path(style).gsub(%r{\A/}, '')}"
```
Comment (princemaple) [inline lib/paperclip/storage/aliyun.rb]: since you are using `\A`, might just use `sub` instead of `gsub`

**Comment by Martin91** [comment]
@aidistan Thanks very much for your contribution, so sorry that I am too busy to reply your work in time, I have invited you to be collaborator of this repo, so feel free to append your name to the authors list at https://github.com/Martin91/paperclip-storage-aliyun/blob/master/paperclip-storage-aliyun.gemspec#L9 and then you can merge this pull request yourself.

**Comment by aidistan** [comment]
@princemaple Thanks for the review. I will fix them in the following commits.

**Comment by aidistan** [comment]
@Martin91 :ok:

**Comment by Martin91** [comment]
@aidistan Additionally, I think I will release a new version `0.1.0` because larger changes from your previous contributions. I plan to finish it by this weekend.

**Comment by aidistan** [comment]
@Martin91 Looking forward to that.

**Pull request merged**